### ISSUE TITLE
Cannot access doc in post save error handling middleware

### ISSUE THREAD

**Issue by makrandgupta** [Automattic/mongoose#7832]
**Do you want to request a *feature* or report a *bug*?**
Reporting a bug

**What is the current behavior?**
In post save error handling middleware, the second argument (expected to contain the doc) is empty.

More context:
I have a field called `numericId` that is unique I am manually incrementing. I have a scenario where there is a race condition which results in a duplicate key error. When I try to save a document with a repeated `numericId`, I want to catch the error in the post save middleware and handle it. 

**If the current behavior is a bug, please provide the steps to reproduce.**
Try to save a document with an existing unique key

My middleware code:
```js
Schema.post('save', function(error, doc, next) {
  if (error.name === 'MongoError' && error.code === 11000) {
		console.log('doc', arguments);
		next(new Error('There was a duplicate key error'));
  } else {
    next();
  }
});
```

The output from the `console.log` statement:
```
doc { '0':
   { MongoError: E11000 duplicate key error collection: test.transactions index: numericId_1 dup key: { : 1005 }
    at Function.create (C:\Users\Makrand\dev\swiftseat-api\node_modules\mongodb-core\lib\error.js:43:12)
    at toError (C:\Users\Makrand\dev\swiftseat-api\node_modules\mongodb\lib\utils.js:149:22)
    at coll.s.topology.insert (C:\Users\Makrand\dev\swiftseat-api\node_modules\mongodb\lib\operations\collection_ops.js:859:39)
    at handler (C:\Users\Makrand\dev\swiftseat-api\node_modules\mongodb-core\lib\topologies\replset.js:1155:22)
    at C:\Users\Makrand\dev\swiftseat-api\node_modules\mongodb-core\lib\connection\pool.js:397:18
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
     driver: true,
     name: 'MongoError',
     index: 0,
     code: 11000,
     errmsg: 'E11000 duplicate key error collection: test.transactions index: numericId_1 dup key: { : 1005 }',
     [Symbol(mongoErrorContextSymbol)]: {} },
  '1': null,
  '2': [Function] }
```
*Note:* The second argument (key = 1) is null. 
**What is the expected behavior?**
The second argument (key = 1) should contain the document I was trying to save.
**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node: 8.11.2
Mongoose: 5.5.10
MongoDB: 4.0

I looked at #5464 and it sounds like its similar but I'm not too sure.

**Comment by vkarpov15** [comment]
Can you please show how you are calling `save()`?

**Comment by makrandgupta** [comment]
We have a bunch of blocking operations that we queue up in an array and then await on that array of promises. 

```js
promises.push(doc.save().catch(err => {
  log.warn(`${tag} Saving doc failed`, err.message, doc);
}));
await Promise.all(promises);
```

**Comment by vkarpov15** [comment]
Thanks for your patience. Fix will be in 5.6.2 :+1:
